decryptor/postgresql/prepared_statements.go

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2020, Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package postgresql
+
+import (
+	"errors"
+
+	"github.com/cossacklabs/acra/decryptor/base"
+	"github.com/cossacklabs/acra/sqlparser"
+)
+
+// Errors returned by prepared statement registry.
+var (
+	ErrStatementNotFound = errors.New("no prepared statement with given name")
+	ErrCursorNotFound    = errors.New("no cursor with given name")
+)
+
+// PgPreparedStatementRegistry is a PostgreSQL PreparedStatementRegistry.
+type PgPreparedStatementRegistry struct {
+	registry map[string]base.PreparedStatement
+}
+
+// NewPreparedStatementRegistry makes a new empty prepared statement registry.
+func NewPreparedStatementRegistry() *PgPreparedStatementRegistry {
+	return &PgPreparedStatementRegistry{registry: make(map[string]base.PreparedStatement)}
+}
+
+// StatementByName returns a prepared statement from the registry by its name, if it exists.
+func (r *PgPreparedStatementRegistry) StatementByName(name string) (base.PreparedStatement, error) {
+	s, ok := r.registry[name]
+	if ok {
+		return s, nil
+	}
+	return nil, ErrStatementNotFound
+}
+
+// Add a statement to the registry. If an existing statement with the same name exists,
+// it is replaced with the new one. Returns "true" if statement has been replaced.
+func (r *PgPreparedStatementRegistry) Add(statement base.PreparedStatement) (bool, error) {
+	name := statement.Name()
+	_, exists := r.registry[name]
+	r.registry[name] = statement
+	return exists, nil
+}
+
+// PgPreparedStatement is a PostgreSQL PreparedStatement.
+type PgPreparedStatement struct {
+	name string
+	text string
```

Comment:
what about to use [OnQueryObject](https://github.com/cossacklabs/acra/blob/master/decryptor/base/observers.go#L27) which cache statement (to avoid one more parsing) and re-generate query on demand

Reply:
Hm... Does it result in much memory saved? I mean, when we're going to created prepared statement objects we will have the SQL text on hand (from the user data), and we'd need to parse it anyway. They don't seem like massive pieces of data so that we can't keep them both around.

Though, I'm not really sure that we will need both text and the parsed query. I guess, for processing we will be interested in the parsed form, and the text might get useful only for debug logging or something. In that case it kinda makes sense to keep only the parsed form and restore the text from it if necessary.

OnQueryObjects will be used later for processing. I intend them to just refer to existing prepared statement from the registry, not keep a personal copy of the parsed query.

Reply:
Main goal of adding OnQueryObject was to avoid extra cpu-computations on parsing string every time. And first time it was introduced for censor + encryption processors. Some censors rules work over parsed query, some of them over sql string.
Then it passed to encryption processors which may change query itself and do it over parsed query. And string representation used only on final stage when statement converted to string to place it into output packet to database. Second goal was to avoid synchronization between statement struct and sql query. If each processor accept string + statement, it doesn't know what will used of these in next processor. So if it updated statement, it should update query string and pass it forward.

Probably we should fill our registry by the same flow of processing query. Create some subscriber for `OnQuery` event (which used by other encryption processors) and be sure that it last processor which will accept OnQueryObject and save in registry after all manipulations by previous processors (instead placing it as separate block of code before generating final presentation of packet to database with updated query)